sync/syncer.go

```diff
@@ -189,24 +195,28 @@ func (s *Syncer) lastTickedLayer() types.LayerID {
 }
 
 func (s *Syncer) Synchronise() {
-	for currentSyncLayer := s.lValidator.ValidatedLayer() + 1; currentSyncLayer <= s.lastTickedLayer(); currentSyncLayer++ {
-		lastTickedLayer := s.lastTickedLayer()
-		s.Info("syncing layer %v current consensus layer is %d", currentSyncLayer, lastTickedLayer)
-
-		if s.IsSynced() && currentSyncLayer == lastTickedLayer {
-			continue
-		}
-
+	defer s.syncFuncWg.Done()
+	currentSyncLayer := s.lValidator.ValidatedLayer() + 1
+	if s.IsSynced() {
 		lyr, err := s.GetLayer(types.LayerID(currentSyncLayer))
 		if err != nil {
-			s.Info("layer %v is not in the database", currentSyncLayer)
-			if lyr, err = s.getLayerFromNeighbors(currentSyncLayer); err != nil {
-				s.Info("could not get layer %v from neighbors %v", currentSyncLayer, err)
-				return
-			}
+			s.Error("err not nil err=%v", err)
+			return
 		}
-
 		s.lValidator.ValidateLayer(lyr) // wait for layer validation
+		return
+	}
+	for ; currentSyncLayer <= s.lastTickedLayer(); currentSyncLayer++ {
```

Comment:
There's an edge case not addressed here:
if a miner goes down (after seeing blocks) in layer j and comes back in (the middle of) layer j+1, she would be considered synced. 
This will cause the following:
1. The miner would not fetch the blocks of layer j+1, when she comes back.
2. She would be considered synced for Hare. Hence, with the current code, she will panic in the end of the protocol, since she won't have the output blocks of the protocol.
3. (ignoring (2)) the miner will be considered not synced at the beginning of layer j+2. She then will fetch layer j+1 and layer j+2, but there is a time race with her neighbours adding blocks of layer j+2 to their database.

Reply:
I opened a bug